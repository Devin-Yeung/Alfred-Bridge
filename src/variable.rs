use serde::Serialize;
use std::collections::HashMap;
use typed_builder::TypedBuilder;

#[repr(transparent)]
#[derive(TypedBuilder)]
#[builder(
    doc,
    field_defaults(default, setter(into)),
    mutators(
        pub fn var(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
            self.variables.insert(key.into(), value.into());
            self
        }
    )
)]

pub struct Variable {
    #[builder(via_mutators)]
    variables: HashMap<String, String>,
}

impl Serialize for Variable {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.variables.serialize(serializer)
    }
}
