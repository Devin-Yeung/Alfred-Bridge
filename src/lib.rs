#![doc(
    html_logo_url = "https://raw.githubusercontent.com/Devin-Yeung/alfred-bridge/master/docs/images/alfred-bridge-logo-square.png"
)]

use crate::variable::Variable;
use serde::Serialize;
use typed_builder::TypedBuilder;

pub mod item;
mod variable;

#[derive(Serialize, TypedBuilder)]
#[builder(
    doc,
    field_defaults(default),
    mutators(
        pub fn item(&mut self, item: item::AlfredItem) -> &mut Self {
            self.items.push(item);
            self
        }
    )
)]
pub struct AlfredOutput {
    #[builder(via_mutators)]
    items: Vec<item::AlfredItem>,
    #[builder(setter(strip_option))]
    #[serde(skip_serializing_if = "Option::is_none")]
    variables: Option<Variable>,
}

impl AlfredOutput {
    pub fn stdout(&self) -> Result<(), serde_json::Error> {
        println!("{}", serde_json::to_string(&self)?);
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_output() {
        let output = AlfredOutput::builder()
            .item(
                item::AlfredItem::builder()
                    .title("Hello, World!")
                    .subtitle("This is a test subtitle.")
                    .arg("https://example.com")
                    .icon(
                        item::Icon::builder()
                            .r#type(item::IconType::FileIcon)
                            .path("icon.png")
                            .build(),
                    )
                    .build(),
            )
            .variables(
                Variable::builder()
                    .var("key", "value")
                    .var("key2", "value2")
                    .build(),
            )
            .build();

        output.stdout().unwrap(); // This will print the JSON to stdout
        insta::assert_json_snapshot!(output)
    }
}
