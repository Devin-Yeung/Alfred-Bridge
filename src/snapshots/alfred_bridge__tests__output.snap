---
source: src/lib.rs
expression: output
---
{
  "items": [
    {
      "title": "Hello, World!",
      "subtitle": "This is a test subtitle.",
      "arg": [
        "https://example.com"
      ],
      "icon": {
        "type": "fileicon",
        "path": "icon.png"
      },
      "valid": true,
      "type": "default"
    }
  ],
  "variables": {
    "key2": "value2",
    "key": "value"
  }
}
